### PR TITLE
Add: List recently executed commands in crashlog output.

### PR DESCRIPTION
**Pull request by JGRennison**
This maintains a circular buffer of info on recent commands, and adds a section to the crash log.

This is useful for providing context and further information of what was occurring shortly prior to a crash.

### CODE REVIEW

**Comment by nielsmh** [comment]
I don't see any way for the CLEF_CMD_FAILED or CLEF_GENERATING_WORLD flags to become set?

**Comment by Eddi-z** [comment]
Could this leak passwords or other personal data?

**Comment by nielsmh** [comment]
Company passwords aren't game commands, but part of the network protocol itself. Server connection passwords aren't game commands either. The command log could contain any string entered for things like vehicle names, group names, signs, etc., i.e. nothing that wouldn't also be stored in the savegame.

**Comment by JGRennison** [comment]
CLEF_CMD_FAILED and CLEF_GENERATING_WORLD are handled at src/command.cpp:398-399.

This does not store text/strings, as these are rarely of any use for debugging. It only stores a flag indicating the presence of a non-empty text parameter.

Passwords and (private) personal data should not go via DoCommandP calls as these are replicated to all clients in multiplayer.

File: src/command.cpp [hunk 1]
```diff
@@ -360,6 +362,85 @@ static const Command _command_proc_table[] = {
...
+	CLEF_NONE                = 0x00, ///< no flag is set
+	CLEF_CMD_FAILED          = 0x01, ///< command failed
+	CLEF_GENERATING_WORLD    = 0x02, ///< generating world
+	CLEF_TEXT                = 0x04, ///< have command text
+	CLEF_ESTIMATE_ONLY       = 0x08, ///< estimate only
+	CLEF_ONLY_SENDING        = 0x10, ///< only sending
+	CLEF_MY_CMD              = 0x20, ///< locally generated command
+};
+DECLARE_ENUM_AS_BIT_SET(CommandLogEntryFlagEnum)
+typedef SimpleTinyEnumT<CommandLogEntryFlagEnum, byte> CommandLogEntryFlag;
```
Comment (nielsmh) [inline src/command.cpp:378]: I think C++11 enum base type declarations are acceptable to use now, i.e. declaring `enum: CommandLogEntryFlag : byte { CLEF_NONE ... };` so separate enum and storage types are not required.

Comment (PeterN) [reply src/command.cpp:378]: I also wonder if we should start using enum classes.

**Comment by nielsmh** [comment]
@JGRennison How useful is this actually in resolving issues? Any examples where this has been the primary lead?

**Comment by TrueBrain** [comment]
Removed "backport requested" as I think we should not add new code to RCs. This could potentially do more harm than good. Lets run this in 1.10, and bring it in from there.

**Comment by JGRennison** [comment]
@nielsmh It's been quite useful in debugging desyncs and crashes/assertion failures involving map or data structure corruption. What commands were recently executed usually provides a big clue as to the cause of the issue.
Conceptually it is similar to the recent news section, though I've been unable to find a use for that so far.

Generally when I receive a crash log I look at: the crash reason lines, scope logging lines, which thread it is, the decoded stack traces, which OS/compiler, AI/company config, the command log and if relevant the game date and lang, blitter, font config. The rest of it is rarely of any use.

**Comment by PeterN** [comment]
Is there a way to make this log AI commands as well?

**Comment by glx22** [comment]
AI commands are as human commands usually, so they are probably already included in the log.

File: src/command.cpp [hunk 2]
```diff
@@ -360,6 +362,85 @@ static const Command _command_proc_table[] = {
...
+	CommandLogEntryFlag log_flags;
+
+	CommandLogEntry() { }
+
+	CommandLogEntry(const CommandCost &res, TileIndex tile, uint32 p1, uint32 p2, uint32 cmd, CommandLogEntryFlag log_flags)
+			: tile(tile), p1(p1), p2(p2), cmd(cmd), date(_date), date_fract(_date_fract),
+			current_company(_current_company), local_company(_local_company), log_flags(log_flags)
+	{
+		if (res.Failed()) this->log_flags |= CLEF_CMD_FAILED;
+		if (_generating_world) this->log_flags |= CLEF_GENERATING_WORLD;
```
Comment (TrueBrain) [inline src/command.cpp]: Wouldn't it be more clear (possibly less error prune) to set all the flags in one place? Either here (and add 4 parameters) or lower? It feels a bit weird now, with this split.
Not sure, mainly asking. I noticed others also asked about this, and it confused me too :D

**Comment by PeterN** [comment]
@glx I thought so but my crashlog only contained AI startup control commands

**Comment by PeterN** [comment]
Ok, AI uses `DoCommandPInternal()` directly, so bypasses the logging in `DoCommandP()`

File: src/command.cpp [hunk 3]
```diff
@@ -600,6 +681,16 @@ bool DoCommandP(TileIndex tile, uint32 p1, uint32 p2, uint32 cmd, CommandCallbac
...
+	CommandLogEntryFlag log_flags;
+	log_flags = CLEF_NONE;
+	if (!StrEmpty(text)) log_flags |= CLEF_TEXT;
+	if (estimate_only) log_flags |= CLEF_ESTIMATE_ONLY;
+	if (only_sending) log_flags |= CLEF_ONLY_SENDING;
+	if (my_cmd) log_flags |= CLEF_MY_CMD;
+	command_log[command_log_next] = CommandLogEntry(res, tile, p1, p2, cmd, log_flags);
+	command_log_next = (command_log_next + 1) % command_log.size();
+	command_log_count++;
+
```
Comment (PeterN) [inline src/command.cpp]: Can you please move this section to `DoCommandPInternal()` so that AI commands are logged as well?

**Review by PeterN — DISMISSED** [review]

**Comment by JGRennison** [comment]
The flags are now all set in one place.

I've moved the logging point to be within DoCommandPInternal, this does change the semantics a bit.
As DoCommandPInternal has multiple return paths, the early error paths where the test command is never executed aren't logged any more, and the logging takes place after the test command is done, before any local command execution with DC_EXEC.
Is this reasonable, or should the logging entry be made/updated after any DC_EXEC execution (in the edge case where the results differ)?
DoCommandP's only_sending flags isn't saved any more, but that is probably no great loss.

File: src/command.cpp (hunk 1 above)
Comment (nielsmh) [reply src/command.cpp:378]: Definitely change this to enum with underlying type ahead of #7538

**Review by nielsmh — CHANGES_REQUESTED** [review]

**Comment by LordAro** [comment]
Hi, there's been no activity on this for nearly 4 months. If this PR is not updated in the next week or so, it will be closed
